include/nil/actor/math/domains/detail/basic_radix2_domain_aux.hpp

```diff
@@ -48,78 +50,77 @@ namespace nil {
-                                        elt *= omega_step;
+                    std::size_t m = 1;    // invariant: m = 2^{s-1}
+                    field_value_type w_m;
+
+                    for (std::size_t s = 1; s <= logn; ++s) {
+                        // w_m is 2^s-th root of unity now
+                        w_m = omega.pow(n / (2 * m));
+                        size_t count_k = n / (2 * m) + (n % (2 * m) ? 1 : 0);
+
+                        if (m > count_k) {
+                            for (std::size_t k = 0; k < n; k += 2 * m) {
+                                // We can parallelize here, because for each range, we touch only 
+                                // a[k + j] and a[k + j + m], and they are never equal for different values of j.
+                                detail::block_execution(
+                                    m,
+                                    smp::count,
+                                    [&a, k, m, &w_m](std::size_t begin, std::size_t end) {
+                                        field_value_type w = w_m.pow(begin);
+                                        for (std::size_t j = begin; j < end; j++) {
+                                            const value_type t = a[k + j + m] * w;
+                                            a[k + j + m] = a[k + j] - t;
+                                            a[k + j] += t;
+                                            w *= w_m;
+                                        }
+                                    }
+                                ).get();
+                            }
+                        } else {
+                            // Here we can parallelize on the cycle with 'k', because for each value of k
+                            // the ranges of array 'a' used do not intersect.
+                            detail::block_execution(
+                                count_k,
+                                smp::count,
+                                [&a, m, &w_m](std::size_t begin, std::size_t end) {
+                                    for (std::size_t k_index = begin; k_index < end; k_index++) {
+                                        field_value_type w = field_value_type::one();
+                                        std::size_t k = k_index * 2 * m;
+                                        size_t w_power = 0;
+                                        for (std::size_t j = 0; j < m; ++j) {
+                                            const value_type t = a[k + j + m] * w;
+                                            a[k + j + m] = a[k + j] - t;
+                                            a[k + j] += t;
+                                            w *= w_m;
+                                        }
                                     }
-                                    elt = elt * omega_j;
                                 }
+                            ).get();
```

Comment:
Just small question - is it better to parallel internal loop and not  for (std::size_t s = 1; s <= logn; ++s) ?

Reply:
We can not parallelize the external loop, they are actively changing the same Range &a, which is an in-out parameter. They will interfere with each other.